app/code/Magento/InventoryInStorePickup/Model/Carrier/Command/GetFreeBoxes.php

```diff
@@ -0,0 +1,56 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryInStorePickup\Model\Carrier\Command;
+
+/**
+ * Calculate and return number of items with free delivery.
+ */
+class GetFreeBoxes
+{
+    /**
+     * @param \Magento\Quote\Model\Quote\Address\RateRequest $request
+     *
+     * @return float
+     */
+    public function execute(\Magento\Quote\Model\Quote\Address\RateRequest $request): float
+    {
+        $freeBoxes = 0.0;
+        if ($request->getAllItems()) {
+            /** @var \Magento\Quote\Model\Quote\Item\AbstractItem $item */
+            foreach ($request->getAllItems() as $item) {
+                if ($item->getProduct()->isVirtual() || $item->getParentItem()) {
```

Comment:
Would be great to avoid breaking the Law of Demeter if those older APIs would allow it.

Reply:
I am not sure if it was a good idea, because we create new dependency on `Magento_Catalog` but please check changes in my last comment.